npm postinstall fails in Windows: rd /s /q worker\out\msys => The system cannot find the file specified

## Bug Report

This is a recent error happening in Windows CI: https://github.com/versatica/mediasoup/actions/runs/6536142903/job/17747178461?pr=1178

```
npm-scripts [INFO] [postinstall] executeCmd(): rd /s /q worker\out\msys
The system cannot find the file specified.
npm-scripts [ERROR] [postinstall] executeCmd() failed, exiting: Error: Command failed: rd /s /q worker\out\msys
```

Nothing has changed so this is related to Windows. No idea. We need to make it work reliably in all supported Windows versions.

Somehow the issue doesn't make sense. When installing mediasoup on Windows (and forcing it to compile the worker instead of fetching a prebuilt worker), this is the flow:

1. `MEDIASOUP_SKIP_WORKER_PREBUILT_DOWNLOAD=true npm install mediasoup --foreground-scripts` (well, whichever syntax is needed in Windows to set environment variables).
2. The npm 'postinstall' task will run the 'postinstall' section in `npm-scripts.mjs`.
3. It will first fun `buildWorker()`:
  ```ts
  function buildWorker()
  {
	logInfo('buildWorker()');

	if (IS_WINDOWS)
	{
		if (!fs.existsSync('worker/out/msys/bin/make.exe'))
		{
			installMsysMake();
		}

		const msysPath = `${process.cwd()}\\worker\\out\\msys\\bin`;

		if (!process.env.PATH.includes(msysPath))
		{
			process.env.PATH = `${msysPath};${process.env.PATH}`;
		}
	}

	executeCmd(`${MAKE} -C worker`);
  }
  ```
4. Note that if `worker/out/msys/bin/make.exe` doesn't exist then it calls `installMsysMake()` which does this:
  ```ts
  function installMsysMake()
  {
	logInfo('installMsysMake()');

	let res = spawnSync('where', [ 'python3.exe' ]);

	if (res.status !== 0)
	{
		res = spawnSync('where', [ 'python.exe' ]);

		if (res.status !== 0)
		{
			logError('`installMsysMake() | cannot find Python executable');

			exitWithError();
		}
	}

	executeCmd(`${String(res.stdout).trim()} worker\\scripts\\getmake.py`);
  }
  ```
5. So here it ensures that we have `python3.exe` and then runs `worker/scripts/getmake.py` which does this:
  ```py
  import io, hashlib, tarfile, urllib.request

  def get(url, digest):
    data = urllib.request.urlopen(url).read()
    assert hashlib.sha256(data).hexdigest() == digest
    tar = tarfile.open(fileobj=io.BytesIO(data))
    tar.extractall('worker/out/msys')
    tar.close()

  get('https://sourceforge.net/projects/mingw/files/MSYS/Base/msys-core/msys-1.0.19-1/msysCORE-1.0.19-1-msys-1.0.19-bin.tar.xz/download', '8c4157d739a460f85563bc4451e9f1bbd42b13c4f63770d43b9f45a781f07858')
  get('https://sourceforge.net/projects/mingw/files/MSYS/Base/libiconv/libiconv-1.14-1/libiconv-1.14-1-msys-1.0.17-dll-2.tar.lzma/download', '196921e8c232259c8e6a6852b9ee8d9ab2d29a91419f0c8dc27ba6f034231683')
  get('https://sourceforge.net/projects/mingw/files/MSYS/Base/gettext/gettext-0.18.1.1-1/libintl-0.18.1.1-1-msys-1.0.17-dll-8.tar.lzma/download', '29db8c969661c511fbe2a341ab25c993c5f9c555842a75d6ddbcfa70dec16910')
  get('https://sourceforge.net/projects/mingw/files/MSYS/Base/coreutils/coreutils-5.97-3/coreutils-5.97-3-msys-1.0.13-bin.tar.lzma/download', 'f8c7990416ea16a74ac336dcfe0f596bc46b8724b2d58cf8a3509414220b2366')
  get('https://sourceforge.net/projects/mingw/files/MSYS/Base/regex/regex-1.20090805-2/libregex-1.20090805-2-msys-1.0.13-dll-1.tar.lzma/download', '85dd8c1e27a90675c5f867be57ba7ae2bb55dde8cd2d19f284c896be134bd3d1')
  get('https://sourceforge.net/projects/mingw/files/MSYS/Base/termcap/termcap-0.20050421_1-2/libtermcap-0.20050421_1-2-msys-1.0.13-dll-0.tar.lzma/download', '62b58fe0880f0972fcc84a819265989b02439c1c5185870227bd25f870f7adb6')
  get('https://sourceforge.net/projects/mingw/files/MSYS/Base/bash/bash-3.1.23-1/bash-3.1.23-1-msys-1.0.18-bin.tar.xz/download', '38da5419969ab883058a96322bb0f51434dd4e9f71de09cd4f75b96750944533')
  get('https://sourceforge.net/projects/mingw/files/MSYS/Base/make/make-3.81-3/make-3.81-3-msys-1.0.13-bin.tar.lzma/download', '847f0cbbf07135801c8e67bf692d29b1821e816ad828753c997fa869a9b89988')
  ```
6. Here it **clearly** extracts things in `worker/out/msys` folder, notice the `tar.extractall('worker/out/msys')` line.
7. So then `buildWorker()` continues. It adds `${process.cwd()}\\worker\\out\\msys\\bin` to the `PATH` and runs `make` to compile the worker.
8. And then `buildWorker()` completes so `cleanWorkerArtifacts()` is called which does this:
  ```ts
  function cleanWorkerArtifacts()
  {
	logInfo('cleanWorkerArtifacts()');

	// Clean build artifacts except `mediasoup-worker`.
	executeCmd(`${MAKE} clean-build -C worker`);
	// Clean downloaded dependencies.
	executeCmd(`${MAKE} clean-subprojects -C worker`);
	// Clean PIP/Meson/Ninja.
	executeCmd(`${MAKE} clean-pip -C worker`);

	if (IS_WINDOWS)
	{
		executeCmd('rd /s /q worker\\out\\msys');
	}
  }
  ```
9. And somehow the last command `rd /s /q worker\\out\\msys` fails only in some cases (see https://github.com/versatica/mediasoup/issues/1179#issuecomment-1770463999) with this error:
  ```
  npm-scripts [INFO] [postinstall] executeCmd(): rd /s /q worker\out\msys
  The system cannot find the file specified.
  npm-scripts [ERROR] [postinstall] executeCmd() failed, exiting: Error: Command failed: rd /s /q worker\out\msys
  ```
9. How is it possible? Obviously `worker\out\msys` folder DOES exist, it's guaranteed by steps before.

--- a/npm-scripts.mjs
+++ b/npm-scripts.mjs
@@ -445,7 +445,9 @@ function installMsysMake()
 		pythonPath = String(res.stdout).trim();
 	}
 
-	executeCmd(`${pythonPath} worker\\scripts\\getmake.py`);
+	const dir = path.resolve('worker/out/msys');
+
+	executeCmd(`${pythonPath} worker\\scripts\\getmake.py --dir="${dir}"`);
 }
 
 function ensureDir(dir)

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -1,4 +1,5 @@
 use std::env;
+#[cfg(target_os = "windows")]
 use std::path::Path;
 use std::process::Command;
 

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -19,8 +19,10 @@ fn main() {
     // Force forward slashes on Windows too so that is plays well with our dumb `Makefile`.
     let mediasoup_out_dir = format!("{}/out", out_dir.replace('\\', "/"));
 
-    // Store original PATH so we make `make clean-all` use it, otherwise
-    // our modified PATH may include rm.exe in Windows and delete itself.
+    // Store original PATH so we make `make clean-all` use it. This is because, in Windows,
+    // we may need to fetch `make` and we store it in out/msys and then we add out/msys/bin
+    // to the PATH, and that folder may contain rm.exe, so `make clean-all` would use
+    // that rm.exe and delete itself and make the task fail.
     let original_path = env::var("PATH").unwrap();
 
     // Add C++ std lib

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -1,6 +1,4 @@
 use std::env;
-#[cfg(target_os = "windows")]
-use std::path::Path;
 use std::process::Command;
 
 fn main() {
@@ -94,7 +92,7 @@ fn main() {
             };
 
             let worker_abs_path = env::current_dir().unwrap();
-            let dir = Path::new(&worker_abs_path).join("out/msys");
+            let dir = worker_abs_path.join("out/msys");
 
             if !Command::new(python)
                 .arg("scripts\\getmake.py")

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -16,7 +16,7 @@ fn main() {
     };
 
     let out_dir = env::var("OUT_DIR").unwrap();
-    // Force forward slashes on Windows too so that is plays well with our dumb `Makefile`
+    // Force forward slashes on Windows too so that is plays well with our dumb `Makefile`.
     let mediasoup_out_dir = format!("{}/out", out_dir.replace('\\', "/"));
 
     // Add C++ std lib
@@ -91,8 +91,7 @@ fn main() {
                 "python".to_string()
             };
 
-            let worker_abs_path = env::current_dir().unwrap();
-            let dir = worker_abs_path.join("out/msys");
+            let dir = format!("{}/msys", mediasoup_out_dir.replace('\\', "/"));
 
             if !Command::new(python)
                 .arg("scripts\\getmake.py")
@@ -104,6 +103,11 @@ fn main() {
             {
                 panic!("Failed to install MSYS/make")
             }
+
+            env::set_var(
+                "PATH",
+                format!("{}\\bin;{}", dir, env::var("PATH").unwrap()),
+            );
         }
 
         env::set_var(

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -19,6 +19,10 @@ fn main() {
     // Force forward slashes on Windows too so that is plays well with our dumb `Makefile`.
     let mediasoup_out_dir = format!("{}/out", out_dir.replace('\\', "/"));
 
+    // Store original PATH so we make `make clean-all` use it, otherwise
+    // our modified PATH may include rm.exe in Windows and delete itself.
+    let original_path = env::var("PATH").unwrap();
+
     // Add C++ std lib
     #[cfg(target_os = "linux")]
     {
@@ -172,6 +176,7 @@ fn main() {
         // Clean
         if !Command::new("make")
             .arg("clean-all")
+            .env("PATH", original_path)
             .env("MEDIASOUP_OUT_DIR", &mediasoup_out_dir)
             .spawn()
             .expect("Failed to start")

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -79,7 +79,7 @@ fn main() {
     {
         if !std::path::Path::new("worker/out/msys/bin/make.exe").exists() {
             let python = if env::var("PYTHON").is_ok() {
-                env::var("PYTHON").unwrap().to_string()
+                env::var("PYTHON").unwrap()
             } else if Command::new("where")
                 .arg("python3")
                 .status()
@@ -92,10 +92,10 @@ fn main() {
             };
 
             let worker_abs_path = env::current_dir().unwrap();
-            let dir = format!("{}\\out\\msys", worker_abs_path.display());
+            let dir = Path::new(&worker_abs_path).join("out/msys");
 
             if !Command::new(python)
-                .arg("scripts\\getmake.py --dir {}", dir)
+                .arg("scripts\\getmake.py --dir {}", dir.display())
                 .status()
                 .expect("Failed to start")
                 .success()

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -1,4 +1,5 @@
 use std::env;
+use std::path::Path;
 use std::process::Command;
 
 fn main() {
@@ -95,7 +96,9 @@ fn main() {
             let dir = Path::new(&worker_abs_path).join("out/msys");
 
             if !Command::new(python)
-                .arg("scripts\\getmake.py --dir {}", dir.display())
+                .arg("scripts\\getmake.py")
+                .arg("--dir")
+                .arg(dir)
                 .status()
                 .expect("Failed to start")
                 .success()

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -96,7 +96,7 @@ fn main() {
             if !Command::new(python)
                 .arg("scripts\\getmake.py")
                 .arg("--dir")
-                .arg(dir)
+                .arg(dir.clone())
                 .status()
                 .expect("Failed to start")
                 .success()

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -78,19 +78,24 @@ fn main() {
     #[cfg(target_os = "windows")]
     {
         if !std::path::Path::new("worker/out/msys/bin/make.exe").exists() {
-            let python = if Command::new("where")
-                .arg("python3.exe")
+            let python = if env::var("PYTHON").is_ok() {
+                env::var("PYTHON").unwrap().to_string()
+            } else if Command::new("where")
+                .arg("python3")
                 .status()
                 .expect("Failed to start")
                 .success()
             {
-                "python3"
+                "python3".to_string()
             } else {
-                "python"
+                "python".to_string()
             };
 
+            let worker_abs_path = env::current_dir().unwrap();
+            let dir = format!("{}\\out\\msys", worker_abs_path.display());
+
             if !Command::new(python)
-                .arg("scripts\\getmake.py")
+                .arg("scripts\\getmake.py --dir {}", dir)
                 .status()
                 .expect("Failed to start")
                 .success()

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -80,8 +80,8 @@ fn main() {
     #[cfg(target_os = "windows")]
     {
         if !std::path::Path::new("worker/out/msys/bin/make.exe").exists() {
-            let python = if env::var("PYTHON").is_ok() {
-                env::var("PYTHON").unwrap()
+            let python = if let Ok(python) = env::var("PYTHON") {
+                python
             } else if Command::new("where")
                 .arg("python3")
                 .status()

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -97,7 +97,7 @@ fn main() {
                 "python".to_string()
             };
 
-            let dir = format!("{}/msys", mediasoup_out_dir.replace("\\", "/"));
+            let dir = format!("{}/msys", mediasoup_out_dir.replace('\\', "/"));
 
             if !Command::new(python)
                 .arg("scripts\\getmake.py")

--- a/worker/build.rs
+++ b/worker/build.rs
@@ -97,7 +97,7 @@ fn main() {
                 "python".to_string()
             };
 
-            let dir = format!("{}/msys", mediasoup_out_dir.replace('\\', "/"));
+            let dir = format!("{}/msys", mediasoup_out_dir.replace("\\", "/"));
 
             if !Command::new(python)
                 .arg("scripts\\getmake.py")

--- a/worker/scripts/getmake.py
+++ b/worker/scripts/getmake.py
@@ -1,10 +1,21 @@
-import io, hashlib, tarfile, urllib.request
+import argparse, io, hashlib, tarfile, urllib.request
+
+argParser = argparse.ArgumentParser()
+
+argParser.add_argument(
+  '--dir',
+  type=str,
+  required=True,
+  help='absolute path of the directoy in which fetched content will be placed'
+)
+
+args = argParser.parse_args()
 
 def get(url, digest):
   data = urllib.request.urlopen(url).read()
   assert hashlib.sha256(data).hexdigest() == digest
   tar = tarfile.open(fileobj=io.BytesIO(data))
-  tar.extractall('worker/out/msys')
+  tar.extractall(args.dir)
   tar.close()
 
 get('https://sourceforge.net/projects/mingw/files/MSYS/Base/msys-core/msys-1.0.19-1/msysCORE-1.0.19-1-msys-1.0.19-bin.tar.xz/download', '8c4157d739a460f85563bc4451e9f1bbd42b13c4f63770d43b9f45a781f07858')

